things/policies/service.go

```diff
@@ -123,18 +124,65 @@ func (svc service) AddPolicy(ctx context.Context, token string, p Policy) (Polic
 
 	// If the client is admin, add the policy
 	if err := svc.checkAdmin(ctx, userID); err == nil {
+		if err := svc.checkSubject(ctx, userID, external, p); err != nil {
+			return Policy{}, err
+		}
+
 		return svc.policies.Save(ctx, p)
 	}
 
 	// If the client has `g_add` action on the object or is the owner of the object, add the policy
 	ar := AccessRequest{Subject: userID, Object: p.Object, Action: "g_add"}
 	if _, err := svc.policies.EvaluateGroupAccess(ctx, ar); err == nil {
+		if err := svc.checkSubject(ctx, userID, external, p); err != nil {
+			return Policy{}, err
+		}
+
 		return svc.policies.Save(ctx, p)
 	}
 
 	return Policy{}, errors.ErrAuthorization
 }
 
+// checkSubject is used to check if the subject is valid.
+//
+// 1. If the subject is a thing (external is false), check the following:
+//   - The user is an admin
+//   - The user is the owner of the thing
+//   - The user has `c_share` action on the thing
+//
+// 2. If the subject is a user (external is true), check the following:
+//   - The user is an admin
+//   - The user is the owner of the user
+//   - The user has `c_share` action on the user
+func (svc service) checkSubject(ctx context.Context, userID string, external bool, p Policy) error {
+	switch external {
+	case false:
+		if err := svc.checkAdmin(ctx, userID); err == nil {
```

Comment:
We have already checked admin in `AddPolicy`.

Reply:
Change the approach to 
- When the user adding the token is admin, we don't check for the subject or object
- When the user is not admin, we check if they have rights to the subject and also object

Reply:
That's the correct approach.